Fix arm toolchains, allow proper aarch64 support

yeswalrus opened the pull request:
Previously all arm builds had BUILD_64 set, regardless of if it was 32 or 64, and SIZEOF_VOID_P was unset.
This removes some duplicate prefix code, properly uses aarch64 for the system processor on arm64,

Commit:
Fix arm toolchains, allow proper aarch64 support

jdonald commented on standard/toolchain-android32.cmake:
```diff
@@ -1,3 +1,5 @@
 set(ARM_TARGET "armeabi-v7a")
 set(ANDROID_NDK_TOOL_PREFIX arm-linux-androideabi)
+set(CMAKE_SYSTEM_PROCESSOR arm)
+set(CMAKE_SIZEOF_VOID_P 4)
```
afaik CMake normally sets `CMAKE_SIZEOF_VOID_P` via a test-compile. We can force its value and have been doing so, but since the goal here is to do things in more of the standard way is there a way we can let CMake figure it out?

yeswalrus replied:
Good call, these lines aren't needed

jdonald commented on standard/toolchain-arm64.cmake:
```diff
@@ -1,13 +1,9 @@
 set(CMAKE_SYSTEM_NAME Linux)
-set(CMAKE_SYSTEM_PROCESSOR arm)
+set(CMAKE_SYSTEM_PROCESSOR aarch64)
```
Yeah, probably good to opt for `aarch64`. From what I can tell OpenCV and libpng use `aarch64` while assimp uses `arm64`.

Commit:
Fix serious bug in standard_project

CMake searches the root Cmake file for 'project', so using stanard_project as a macro causes it to insert it's own call to project(C CXX) at the very beginning.
CMAKE_SYSTEM_PROCESSOR is set in the toolchain, and so must be queried AFTER the first call to project

Commit:
Set CMAKE_ Variables in post-init unless absolutely nessecary

jdonald commented:
Sorry I haven't gotten to test this before having to run now.

@yeswalrus did you test this with a test-merge into Autowiring (or a similar lib), build that branch via ExternalBuilder, then confirm that ARM Linux 64-bit platform still builds/links and runs some of its tests? That's how I was planning to verify this.

yeswalrus commented:
I did, though locally on my arm64 docker machine. I can kick off a jenkins build

yeswalrus commented:
LeapResource/LeapSerial seems to be having some issues...

jdonald commented on standard/StandardProject.cmake:
```diff
@@ -18,23 +18,33 @@ include(CMakeParseArguments) # Backwards compatibility
 
 # This must be a macro since project defines scope-local variables
 # that we generally rely on being in the root context.
-macro(standard_project project_name)
```
When I test this with a local merge into autowiring, I get the following error :
```
CMake Error at CMakeLists.txt:5 (standard_project):
  Unknown CMake command "standard_project".


-- Configuring incomplete, errors occurred!
```

The problem appears to be the removal of the `standard_project` macro definition on this line, and renaming it to `project` (?)

yeswalrus replied:
Yeah - any reference to standard_project will have to be changed, and any explicit references to preinit and postinit should be too - I found the source of one of the comments in our toolchain file about getting parsed twice.
It turns out that if you don't have a function call to 'project' in your root CMakeLists.txt, cmake will insert a generic call for you at the very beginning of the file, which will mess up some things. Thus, our only option to have this stuff done automatically is to override the project function with our own. 

Take a look at the fix-arm branch if you want to test it out.

yeswalrus commented:
Ok - All the libraries seem to be building ok with these changes. While overriding project would provide somewhat more standard behavior, it does introduce it's own set of challenges, particularly since overriding isn't exactly well documented feature of CMake.

jdonald commented on standard/StandardProject.cmake:
```diff
@@ -18,8 +18,14 @@ include(CMakeParseArguments) # Backwards compatibility
 # that we generally rely on being in the root context.
+# This make use of a little-known feature that allows us to overload builtin functions
+# IMPORTAINT: Note that if there is no *direct* call to 'project' in the root CMakeLists.txt
+# file, CMake will automatically inject one with the default languages of C and CXX.
```
We found this to be not the case, at least in my local tests, so this comment appears stale.

yeswalrus replied:
https://cmake.org/cmake/help/v3.7/command/project.html disagrees.

jdonald commented on standard/toolchain-android.cmake:
```diff
@@ -1,10 +1,8 @@
-# For reasons beyond my comprehension, this file is parsed multiple times
-# and in some of them, cache variables are not preserved so we have to store
-# this in an environmental variable...
+# Deal with wierdnesses caused by our standard_project setup. See
```
Fix spelling.

jdonald commented on standard/StandardProject.cmake:
```diff
@@ -18,8 +18,13 @@ include(CMakeParseArguments) # Backwards compatibility
   cmake_parse_arguments(standard "" "VERSION" "LANGUAGES" ${ARGN})
+
   if(NOT standard_VERSION)
     message(FATAL_ERROR "Standard compliant projects must specify a version")
```
Just noticed this longstanding typo and will fix in another PR...

jdonald replied:
Wait, no demo here. Deleting my comment.

jdonald commented on standard/StandardProject.cmake:
```diff
@@ -18,8 +18,13 @@ include(CMakeParseArguments) # Backwards compatibility
 # This must be a macro since project defines scope-local variables
 # that we generally rely on being in the root context.
+# IMPORTAINT: Note that if there is no *direct* call to 'project' in the root CMakeLists.txt
+# file, CMake will automatically inject one with the default languages of C and CXX.
```
Thanks @yeswalrus for clarifying this via the documentation at https://cmake.org/cmake/help/v3.0/command/project.html !

Only typo remaining I see here is `s/IMPORTAINT/IMPORTANT/`

jdonald commented:
(With a test-merge in autowiring) I verified builds on my Mac for Android 32-bit, Android 64-bit, and OS X (confirming it got the right SDK). Any concerns, @gittyupagain ?

jdonald commented:
Strange, Travis isn't running anymore?

jdonald commented:
Oh right, we don't have such a check for `standard`.
